content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,454 @@
+types in its API. Consequently, you can implement logic that automatically registers the resource
+types you need in Kubernetes.
+
+## When not to use the API Aggregation Layer
+
+There are some anti-patterns where using the API Aggregation Layer isn't recommended.
+Let's go through them.
+
+### Unstable backend
+
+If your api-server stops responding for some reason due to an unavailable backend or other issues it
+may block some Kubernetes functionality. For example, when deleting namespaces, Kubernetes will wait
+for a response from your api-server to see if there are any remaining resources.
+If the response doesn't come, the namespace deletion will be blocked.
+
+Also, you might have encountered a [situation](https://github.com/kedacore/keda/issues/4224) where,
+when the metrics-server is unavailable, an extra message appears in stderr after every API request
+(even unrelated to metrics) stating that `metrics.k8s.io` is unavailable. This is another example
+of how using the API Aggregation Layer can lead to problems when the api-server handling requests
+is unavailable.
+
+### Slow requests
+
+If you can't guarantee an instant response for user requests, it's better to consider using the operator
+pattern. Otherwise, you might make your cluster less stable. Many projects implement an extension
+api-server only for a limited set of resources, particularly for imperative logic and subresources.
+This recommendation is also
+[mentioned in the official Kubernetes](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/#response-latency)
+documentation.
+
+## Why we needed it in Cozystack
+
+As a reminder, we're developing the open-source PaaS platform [Cozystack](https://cozystack.io/),
+which can also be used as a framework for building your own private cloud. Therefore, the ability
+to easily extend the platform is crucial for us.
+
+Cozystack is built on top of [FluxCD](https://fluxcd.io/). Any application is packaged into its
+own Helm chart, ready for deployment in a tenant namespace. Deploying any application on the platform
+is done by creating a `HelmRelease` resource, specifying the chart name and parameters for the application.
+All the rest logic is handled by FluxCD. This pattern allows us to easily extend the platform with new
+applications and provide the ability to create new applications that just need to be packaged
+into the appropriate Helm chart.
+
+{{< figure src="cozystack.png" alt="Interface of the Cozystack platform" caption="Interface of the Cozystack platform" >}}
+
+So, in our platform, everything is configured as `HelmRelease` resources. However, we ran into
+two problems: limitations of the RBAC model and the need for a public API. Let's delve into these
+
+### Limitations of the RBAC model
+
+The standard RBAC system in Kubernetes doesn't allow you to restrict access to a list of resources
```

Comment:
```suggestion
The widely-deployed RBAC system in Kubernetes doesn't allow you to restrict access to a list of resources
```
RBAC is standardized but explicitly not part of conformance. Let's avoid implying otherwise.